enhancements/sig-policy/66-delete-policy-resources/README.md

```diff
@@ -0,0 +1,224 @@
+`spec.removeChildrenOnDelete`, we will set a finalizer on the configurationPolicy,
+such as `policy.open-cluster-management.io/delete-related-objects`. When the policy
+is deleted, the controller will perform cleanup logic on the objects in `relatedObjects`
+that have been marked for deletion by the new status field. Once that cleanup is complete,
+the finalizer will be removed and kubernetes will handle removing the configurationPolicy
+from the cluster. Any resources without this value will persist on the cluster.
+
+Since `removeChildrenOnDelete` and `spec.removeChildrenOnDelete` are scoped to the
+configurationPolicy level, the configurationPolicy controller will handle removing
+objects on each managed cluster once a configurationPolicy is deleted there. This also
+means that when the placement of a policy changes to a different managed cluster
+and that policy contains a configurationPolicy with `removeChildrenOnDelete: true`, the
+object will be deleted on the old cluster and created on the new one. This means that
+changing the placement of an `enforce` policy will no longer result in objects
+persisting indefinitely on the old managed cluster.
+
+
+### Risks and Mitigation
+
+The main risk of adding an enhancement like this would be users deleting objects
+that they want to remain on the cluster after they delete a policy. We plan to
+mitigate this risk by making the `spec.removeChildrenOnDelete` field default to
+`false`, so users will not delete anything without having to set that field first.
+
+## Design Details
+
+The `spec.removeChildrenOnDelete` field would look something like this:
+
+```
+spec:
+  namespaceSelector:
+    exclude:
+    - kube-*
+    include:
+    - default
+  object-templates:
+  - complianceType: musthave
+    objectDefinition:
+      apiVersion: v1
+      kind: Pod
+      metadata:
+        name: proposal-pod
+      spec:
+        containers:
+        - image: nginx:1.18.0
+          name: nginx
+          ports:
+          - containerPort: 80
+  remediationAction: enforce
+  severity: low
+  removeChildrenOnDelete: true //controller will remove proposal-pod
+```
+
+Here is what the configurationPolicy status would look like with the new field:
+
+```
+status:
+  compliancyDetails:
+  - Compliant: Compliant
+    Validity: {}
+    conditions:
+    - lastTransitionTime: "2022-06-09T19:27:13Z"
+      message: pods [demo] in namespace default found as specified, therefore this Object template is compliant
+      reason: K8s `must have` object already exists
+      status: "True"
+      type: notification
+  compliant: Compliant
+  lastEvaluated: "2022-06-09T19:27:53Z"
+  lastEvaluatedGeneration: 1
+  relatedObjects:
+  - compliant: Compliant
+    object:
+      apiVersion: v1
+      kind: pods
+      metadata:
+        name: demo
+        namespace: default
+    properties:
+      createdByParent: true //delete when parent configPolicy is deleted
```

Comment:
what if a user deletes the object then controller creates it, will it be set to true? You probably want to track the uid here since it could happen that the controller deletes one and another controller create a new one with new uid, and you may not want to delete the recreated object.

Reply:
Tracking the uid could be good, although I think it's a weird situation.

It seems like it would only be necessary if the configpolicy creates an object, which is then deleted and recreated *before* the configpolicy tries to create it again. Because in this case the configpolicy didn't create the instance that it's still tracking.

But in general, whenever the configpolicy creates an object as opposed to editing it, I think we want to flag it like this. Regardless of whether that's when the configpolicy was first defined, or whether that happens later (eg after something deletes the object).

Reply:
I'm wondering if we really need to distinguish whether a policy created the object in the first place or not. If a user sets the option to delete the objects after the policy is deleted, I think the objects should just be deleted regardless of who created it and when it was created.

For example, if you manually test out something on a managed cluster and then later make the policy do it on all managed clusters, it seems strange that the object would be deleted on all managed clusters except for the one that was manually tested it on.

Reply:
These seem like additional reasons to consider an enum, we could have options like `AlwaysDelete`, `DeleteIfCreated`.

I think the "if created" logic originally came from wanting to be able to add this flag to existing policies, and do what we imagined users would want. But we could just give that option to the policy authors, maybe even per object-template for more control. We moved away from that because it is necessarily more complex for the user to need to define it in multiple places, but it would make the behavior more clear...